docs/spec_draft.md

```diff
@@ -300,6 +316,148 @@ the IEEE-754 specification. For boolean element type, the behavior is same as
+
+3. If `channel_handle` is provided and `use_global_device_ids` $=$ `true`
+   * Number of groups formed: G.
+   * Size of each group: S.
+   * Interpretation of `replica_groups`: contains global device ids.
+   * The formation of `groups` can be demonstrated as follows:
+     ```python
+     for g in replica_groups:
+       for global_device_id in g:
+           group += ei(global_device_id / P, global_device_id % P)
+       groups += group
+     return groups
+     ```
+    * An example assuming P = 2: `replica_groups`: `[[4, 5], [6, 7]]` and the
+      `groups` formed: `{ei(2, 0), ei(2, 1)} and {ei(3, 0), ei(3, 1)}`.
+
+For each `group` $\in$ `groups`, the `result`, to be visible by each member of
+`group`, is given by `concatenate(operands, all_gather_dim)`, where
+`operands` = { `operand` w.r.t each `ei(rid, pid)` following the same order in
+`group` }.
+
+### Inputs
+
+| Name                    | Type                                                                                                                                         |
+|-------------------------|----------------------------------------------------------------------------------------------------------------------------------------------|
+| `operand`               | tensor of any supported type                                                                                                                 |
+| `all_gather_dim`        | constant of type `si64`                                                                                                                      |
+| `replica_groups`        | 2-dimensional tensor constant of type `si64`                                                                                                  |
+| `channel_handle`        | constant of type struct { `channel-id`: `si64`, `channel-type`: enum of `CHANNEL_TYPE_INVALID`, `DEVICE_TO_DEVICE`, `DEVICE_TO_HOST`, `HOST_TO_DEVICE` } |
+| `use_global_device_ids` | constant of type `boolean`                                                                                                                   |
+
+### Outputs
+
+| Name     | Type                         |
+|----------|------------------------------|
+| `result` | tensor of any supported type |
+
+### Constraints
+
+  * (C1) `all_gather_dim` $\in$ [0, rank(`operand`).
+  * (C2) dim(`operand`, `all_gather_dim`) $\ne$ 0
+  * (C3) dim(`result`, `all_gather_dim`) % dim(`operand`, `all_gather_dim`) $=$ 0.
+  * (C4) On `replica_groups`
+    * All the sub-groups in `replica_groups` are of same size.
+    * The replica ids should cover all the value in the range [0, n), where
+      n is the number of replicas in `replica_groups`.
+    * An empty `replica groups` is allowed only when `use_global_device_ids` is
+      `false`.
+  * (C5) element_type(`operand`) $=$ element_type(`result`).
+  * (C6) rank(`operand`) $=$ rank(`result`).
+  * (C7) dim(`result`, i) is given by
```

Comment:
Feedback:
* disambiguate the order of operations by adding parenthesis around multiply
```
  * (C7) dim(`result`, `i`) is given by:
    * dim(`operand`, `i`), `i` $\ne$ `all_gather_dim`.
    * dim(`result`, `all_gather_dim`) / (dim(`operand`, `all_gather_dim`) * dim(`operand`, `i`)) otherwise.
```